test/qa/test_fileshare.py

```diff
@@ -393,47 +456,112 @@ def test_fileshare_transfer_multiple_files_selective_accept(background: bool):
+
+        transfer_remote = ssh_client.exec_command(f"nordvpn fileshare list {local_transfer_id}")
+        assert fileshare.for_all_files_in_transfer(transfer_remote, canceled_transfer_paths, lambda file_entry: "canceled" in file_entry)
+        assert fileshare.for_all_files_in_transfer(transfer_remote, wfolder_4.transfer_paths, lambda file_entry: "downloaded" in file_entry)
+
+        assert fileshare.files_from_transfer_exist_in_filesystem(local_transfer_id, [wfolder_4], ssh_client)
+
+    time.sleep(1)
 
     transfers = sh.nordvpn.fileshare.list()
     assert "completed" in fileshare.find_transfer_by_id(transfers, peer_transfer_id)
 
-    transfer = ssh_client.exec_command(f"nordvpn fileshare list {peer_transfer_id}")
-    assert "canceled" in fileshare.find_file_in_transfer(wdir.transfer_paths[0], transfer.split("\n"))
-    assert "canceled" in fileshare.find_file_in_transfer(wdir.transfer_paths[1], transfer.split("\n"))
-    assert "downloaded" in fileshare.find_file_in_transfer(wdir.transfer_paths[2], transfer.split("\n"))
-    assert "canceled" in fileshare.find_file_in_transfer(wdir.transfer_paths[3], transfer.split("\n"))
-
     transfers = ssh_client.exec_command("nordvpn fileshare list")
     assert "completed" in fileshare.find_transfer_by_id(transfers, peer_transfer_id)
 
-    assert fileshare.files_from_transfer_exist_in_filesystem(local_transfer_id, [wdir], ssh_client)
-    ssh_client.exec_command(f"sudo rm -rf {workdir}/{os.path.basename(wdir.dir_path)}")
-
     if not background:
         assert fileshare.validate_transfer_progress(command_handle.stdout.decode())
+        assert len(re.findall(fileshare.INTERACTIVE_TRANSFER_PROGRESS_COMPLETED_PATTERN, command_handle.stdout.decode())) == 1
         assert command_handle.is_alive() is False
         assert command_handle.exit_code == 0
 
     assert fileshare.validate_transfer_progress(t_progress_interactive)
+    assert len(re.findall(fileshare.INTERACTIVE_TRANSFER_PROGRESS_COMPLETED_PATTERN, t_progress_interactive)) == 1
 
+    for entity in [wfolder_1, wfolder_2, wfolder_3, wdir_1, wdir_2]:
+        shutil.rmtree(entity.dir_path)
 
-def test_fileshare_graceful_cancel():
-    wdir = fileshare.create_directory(1)
+    ssh_client.exec_command(f"sudo rm -rf {peer_filepath}/*tmp*")
+
+
+@pytest.mark.parametrize("transfer_entity", list(fileshare.FileSystemEntity), ids = [f"send_{entity.value}" for entity in list(fileshare.FileSystemEntity)])
+def test_fileshare_graceful_cancel(transfer_entity: fileshare.FileSystemEntity):
+    wdir = fileshare.create_directory(0)
+    wfolder = fileshare.create_directory(2, parent_dir=wdir.dir_path)
+
+    if transfer_entity == fileshare.FileSystemEntity.FILE:
+        path = wfolder.paths[0]
+        expected_files = [wfolder.filenames[0]]
+    elif transfer_entity == fileshare.FileSystemEntity.FOLDER:
+        wfolder = fileshare.create_directory(2)
```

Comment:
why create a new folder, shouldn't be the same?

Reply:
You might notice this approach being used in a few tests I modified. In my opinion, it's quicker and easier to use `wfolder = fileshare.create_directory(2)` upfront, rather than later creating a new `expected_files` list that'd contain partially stripped `transfer_paths` from `wfolder`. This approach simply avoids extra work, and IMO is less messy :smile: